switchapi/es2k/switch_lag.c

```diff
@@ -0,0 +1,472 @@
+ *   @brief Create LAG
+ *
+ * Arguments:
+ *   @param[in] device - device
+ *   @param[in] api_lag_info - LAG info
+ *   @param[out] lag_h - LAG handle
+ *
+ * Return Values:
+ *    @return  SWITCH_STATUS_SUCCESS on success
+ *             Failure status code on error
+ */
+switch_status_t switch_api_lag_create(switch_device_t device,
+                                      switch_api_lag_info_t* api_lag_info,
+                                      switch_handle_t* lag_h) {
+  switch_status_t status = SWITCH_STATUS_SUCCESS;
+  switch_lag_info_t* lag_info = NULL;
+
+  if (api_lag_info->rmac_handle == SWITCH_API_INVALID_HANDLE) {
+    status = switch_api_device_default_rmac_handle_get(
+        device, &api_lag_info->rmac_handle);
+    CHECK_RET(status != SWITCH_STATUS_SUCCESS, status);
+  }
+
+  if (!SWITCH_RMAC_HANDLE(api_lag_info->rmac_handle)) {
+    status = SWITCH_STATUS_INVALID_HANDLE;
+    krnlmon_log_error(
+        "lag create: Invalid rmac handle on device %d: "
+        "error: %s\n",
+        device, switch_error_to_string(status));
+    return status;
+  }
+
+  *lag_h = switch_lag_handle_create(device);
+  CHECK_RET(*lag_h == SWITCH_API_INVALID_HANDLE, SWITCH_STATUS_NO_MEMORY);
+
+  status = switch_lag_get(device, *lag_h, &lag_info);
+  CHECK_RET(status != SWITCH_STATUS_SUCCESS, status);
+
+  lag_info->lag_handle = *lag_h;
+
+  status = SWITCH_LIST_INIT(&(lag_info->lag_members));
+  SWITCH_ASSERT(status == SWITCH_STATUS_SUCCESS);
+
+  SWITCH_MEMCPY(&lag_info->api_lag_info, api_lag_info,
+                sizeof(switch_api_lag_info_t));
+
+  return status;
+}
+
+/*
+ * Routine Description:
+ *   @brief Delete LAG
+ *
+ * Arguments:
+ *   @param[in] device - device
+ *   @param[out] lag_h - LAG handle
+ *
+ * Return Values:
+ *    @return  SWITCH_STATUS_SUCCESS on success
+ *             Failure status code on error
+ */
+switch_status_t switch_api_lag_delete(switch_device_t device,
+                                      switch_handle_t lag_h) {
+  switch_status_t status = SWITCH_STATUS_SUCCESS;
+  switch_lag_info_t* lag_info = NULL;
+
+  status = switch_lag_get(device, lag_h, &lag_info);
+  CHECK_RET(status != SWITCH_STATUS_SUCCESS, status);
+
+  //--------------------- Tx Path : Del Case ----------------------//
+  status = switch_pd_tx_lag_table_entry(device, lag_info, false);
+  if (status != SWITCH_STATUS_SUCCESS) {
+    krnlmon_log_error(
+        "Failed to delete tx lag table entry on device %d: "
+        ",error: %s\n",
+        device, switch_error_to_string(status));
+    return status;
+  }
+
+  //--------------------- Rx Path : Del Case ----------------------//
+  status = switch_pd_rx_lag_table_entry(device, lag_info, false);
+  if (status != SWITCH_STATUS_SUCCESS) {
+    krnlmon_log_error(
+        "Failed to delete rx lag table entry on device %d: "
+        ",error: %s\n",
+        device, switch_error_to_string(status));
+    return status;
+  }
+
+  status = switch_lag_handle_delete(device, lag_h);
```

Comment:
We aren't deleting the lag handle incase we are not able to delete the the LAG related table entries. We have error handling for the Tx and Rx delete calls, which incase if it fails, the error status will be returned back to the caller without deleting the lag handle. But I do see a case where lets say if the first call(Tx delete call) goes fine, but we hit error in the Rx del path, so we will end up in a situation where the TX entry would not be present in the HW, but the RX entry would be present in the HW and the corresponding lag handle will be present in the software stack, so the HW and SW can go out of sync.

Better approach:
I don't know if we have an infra where we can do some audit, basically poll the HW tables on regular intervals and match it with the SW constructs and delete any stale entries in the HW or in SW if needed. That way we can keep our SW and HW in sync. But I don't think we are doing that as of now. 

So I went with this approach, where at least if the first lag handle goes out of sync, at least for the next lag_handle we will generate new handle and it will be treated as fresh configuration(lag handle is used as key in programming LAG related tables). But on the other hand, if we delete the lag handle, every time it will be allocated the same id and will try to write in the HW with the same key, which will throw "entry already exists" error, because the previous delete didn't happen properly. So in my opinion it is better not the delete the handle if we fail to delete the table entries.

Let me know if this approach works for now, or we can address it in a different way.

Reply:
I like the approach you suggested of polling at regular intervals and deleting stale entries if any but since the infra is not available, the current approach looks good.